internal/http/services/owncloud/ocdav/dav.go

```diff
@@ -156,6 +162,51 @@ func (h *DavHandler) Handler(s *svc) http.Handler {
 	})
 }
 
+func deletePublicShareIfExpired(ctx context.Context, s *svc, token string) error {
+	c, err := pool.GetGatewayServiceClient(s.c.GatewaySvc)
+	if err != nil {
+		return err
+	}
+
+	response, err := c.GetPublicShareByToken(ctx, &linkv1beta1.GetPublicShareByTokenRequest{
+		Token: token,
+	})
+	if err != nil {
+		return err
+	}
+
+	if response.Status.Code != rpc.Code_CODE_OK {
+		return fmt.Errorf("error getting public share")
+	}
+
+	share := response.Share
+	t := time.Unix(int64(share.Expiration.GetSeconds()), int64(share.Expiration.GetNanos()))
+	if share.Expiration != nil && t.Before(time.Now()) {
+		req := &linkv1beta1.RemovePublicShareRequest{
+			Ref: &linkv1beta1.PublicShareReference{
+				Spec: &linkv1beta1.PublicShareReference_Id{
+					Id: &linkv1beta1.PublicShareId{
+						OpaqueId: share.Id.OpaqueId,
+					},
+				},
+			},
+		}
+
+		res, err := c.RemovePublicShare(ctx, req)
+		if err != nil {
+			return err
+		}
+		if res.Status.Code != rpc.Code_CODE_OK {
+			if res.Status.Code == rpc.Code_CODE_NOT_FOUND {
+				return fmt.Errorf("not found")
+			}
+			return fmt.Errorf("grpc delete share request failed")
+		}
+	}
+
+	return nil
```

Comment:
you should also return a bool that tells whether a deletion happened, so that the caller knows whether to continue proceeding or just exit with a 404. because once the share is deleted we should not continue with whatever operation was originally intended